mvrx/src/main/kotlin/com/airbnb/mvrx/CoroutinesStateStore.kt

```diff
@@ -0,0 +1,120 @@
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.selects.select
+import java.util.concurrent.Executors
+
+@Suppress("EXPERIMENTAL_API_USAGE")
+class CoroutinesStateStore<S : MvRxState>(
+    initialState: S,
+    scope: CoroutineScope
+) : MvRxStateStore<S> {
+
+    private val setStateChannel = Channel<S.() -> S>(capacity = Channel.UNLIMITED)
+    private val withStateChannel = Channel<(S) -> Unit>(capacity = Channel.UNLIMITED)
+
+    /**
+     * This combination of BroadcastChannel, mutable state property and channelFlow was arrived
+     * at after multiple attempts. This code needs to:
+     * 1) Multicast
+     * 2) Never fail to deliver an intermediate state to subscribers even if updates are fast or
+     *    some subscribers are slow.
+     * 3) Be able to provide an initial value.
+     *
+     * StateFlow can't be used because it conflates new sets. If you set its value in a tight
+     * loop and listen to changes on a flow, you will miss values. ConflatedBroadcastChannel
+     * alone has the same issue.
+     *
+     * A normal Channel can't be used because it isn't multicast.
+     */
+    private val stateChannel = BroadcastChannel<S>(capacity = Channel.BUFFERED)
+    override var state = initialState
+    override val flow: Flow<S>
+        get() = flow {
+            emit(state)
+            stateChannel.consumeEach { emit(it) }
+        }.buffer(1).distinctUntilChanged()
+
+    init {
+        setupTriggerFlushQueues(scope)
+        scope.coroutineContext[Job]?.invokeOnCompletion {
+            setStateChannel.close()
+            withStateChannel.close()
+        }
+    }
+
+    /**
+     * Observe [flushQueuesChannel] and flush queues whenever there is a new item.
+     * This no-ops if [MvRxTestOverrides.FORCE_SYNCHRONOUS_STATE_STORES] is set.
+     */
+    private fun setupTriggerFlushQueues(scope: CoroutineScope) {
+        if (MvRxTestOverrides.FORCE_SYNCHRONOUS_STATE_STORES) return
+
+        val executor = Executors.newSingleThreadExecutor()
+        scope.coroutineContext[Job]!!.invokeOnCompletion {
+            executor.shutdownNow()
+        }
+
+        scope.launch(executor.asCoroutineDispatcher()) {
+            while (isActive) {
+                flushQueuesOnce()
+            }
+        }
+    }
+
+    /**
+     * Flush the setState and withState queues.
+     * All pending setState reducers will be run prior to every single withState lambda.
+     * This ensures that situations such as the following will work correctly:
+     *
+     * Sitaution 1
+     *
+     * setState { ... }
+     * withState { ... }
+     *
+     * Sitaution 2
+     *
+     * withState {
+     *     setState { ... }
+     *     withState { ... }
+     * }
+     */
+    private suspend fun flushQueuesOnce() {
+        select<Unit> {
+            setStateChannel.onReceive { reducer ->
+                val newState = state.reducer()
+                stateChannel.offer(newState)
+                state = newState
```

Comment:
I'm thinking about checking that `state != newState` here as previously, to use buffer effectively

Reply:
@denis-bezrukov By correctly do you mean that it allows ~64 unique elements to be in the buffer at any given time? Some people's state classes are very large such that hashing it may take a non-trivial amount of time so if it can be limited to a single time (the distinctUntilChanged() on the Flow) that would be ideal.

Reply:
>64 unique elements to be in the buffer at any given time?

yes

>Some people's state classes are very large such that hashing it may take a non-trivial amount of time so if it can be limited to a single time (the distinctUntilChanged() on the Flow) that would be ideal.

since `flow {}` is cold, `distinctUntilChanged()` on it will be called for every consumer

Reply:
Great point!